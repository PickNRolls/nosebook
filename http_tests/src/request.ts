import supertest, { Test } from 'supertest';
import TestAgent from 'supertest/lib/agent';

const HOST = 'http://backend:8080';
const SESSION_HEADER_KEY = 'X-Auth-Session-Id';

class TestAgentWrapper {
  private host: string;
  private agent: TestAgent;

  public constructor(host: string) {
    this.host = host;
    this.agent = supertest(host);
  }

  public unwrap(): TestAgent {
    return this.agent;
  }

  public extend(prefixUrl: string): TestAgentWrapper {
    return new TestAgentWrapper(this.host + prefixUrl);
  }

  private hook<M extends 'get' | 'post'>(method: M, url: string): Test {
    const test = this.agent[method](url);

    test.set(SESSION_HEADER_KEY, 'bb23af03-be50-4bce-b729-b259b2e02e54');

    return test;
  }

  public get(url: string): Test {
    return this.hook('get', url);
  }

  public post(url: string): Test {
    return this.hook('post', url);
  }
};

export default new TestAgentWrapper(HOST);


