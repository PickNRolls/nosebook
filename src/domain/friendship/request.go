package domainfriendship

import (
	"nosebook/src/errors"
	"time"

	"github.com/google/uuid"
)

type FriendRequest struct {
	RequesterId uuid.UUID `json:"requesterId" db:"requester_id"`
	ResponderId uuid.UUID `json:"responderId" db:"responder_id"`
	Message     string    `json:"message" db:"message"`
	Accepted    bool      `json:"accepted" db:"accepted"`
	Viewed      bool      `json:"viewed" db:"viewed"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`

	events []Event
}

func New(
	requesterId uuid.UUID,
	responderId uuid.UUID,
	message string,
	accepted bool,
	viewed bool,
	createdAt time.Time,
	raiseCreatedEvent bool,
) *FriendRequest {
	output := &FriendRequest{
		RequesterId: requesterId,
		ResponderId: responderId,
		Message:     message,
		Accepted:    accepted,
		Viewed:      viewed,
		CreatedAt:   createdAt,
		events:      make([]Event, 0),
	}

	if raiseCreatedEvent {
		output.raiseEvent(NewCreatedEvent())
	}

	return output
}

func (this *FriendRequest) raiseEvent(event Event) {
	this.events = append(this.events, event)
}

func (this *FriendRequest) Events() []Event {
	return this.events
}

func (this *FriendRequest) ViewBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отметить заявку просмотренной")
	}

	this.Viewed = true
	this.raiseEvent(NewViewedEvent())

	return nil
}

func (this *FriendRequest) AcceptBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может принять заявку")
	}

	this.Accepted = true
	this.Viewed = true
	this.raiseEvent(NewAcceptedEvent())

	return nil
}

func (this *FriendRequest) DenyBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отклонить заявку")
	}

	this.Accepted = false
	this.Viewed = true
	this.raiseEvent(NewDeniedEvent())

	return nil
}

func (this *FriendRequest) RemoveBy(userId uuid.UUID) *errors.Error {
	if !this.Accepted {
		return errors.New("Friendship Error", "Нельзя удалить непринятую заявку")
	}

	this.Accepted = false
	previousRequesterId := this.RequesterId
	previousResponderId := this.ResponderId

	if this.RequesterId == userId {
		this.RequesterId, this.ResponderId = this.ResponderId, this.RequesterId
	}

	this.raiseEvent(NewRemovedEvent(previousRequesterId, previousResponderId))

	return nil
}
