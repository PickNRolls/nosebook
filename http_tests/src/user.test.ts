import request from './request';

describe('users', () => {
  let users = request.extend({ prefixUrl: '/users' });

  describe('online', () => {
    test('GET /:id', async () => {
      const response = await users
        .get('/ed1a3fd0-4d0b-4961-b4cd-cf212357740d')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('offline', () => {
    test('GET /:id', async () => {
      const response = await users
        .get('/1ae02f69-ea1a-4308-b825-0e5896e652e4')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('GET /', () => {
    test('fuzzy text', async () => {
      let response = await users
        .get('')
        .query({
          text: 'tester',
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();

      
      response = await users
        .get('')
        .query({
          text: 'drug',
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });
});

