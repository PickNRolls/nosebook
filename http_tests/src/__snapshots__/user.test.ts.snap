// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users offline GET /:id 1`] = `
{
  "data": {
    "firstName": "Ilya",
    "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
    "lastName": "Blinkov",
    "lastOnlineAt": "2024-08-10T10:01:02Z",
    "nick": "drugtester",
    "online": false,
  },
  "ok": true,
}
`;

exports[`users online GET /:id 1`] = `
{
  "data": {
    "firstName": "Test",
    "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
    "lastName": "Tester",
    "lastOnlineAt": "2024-08-10T10:10:10Z",
    "nick": "test_tester",
    "online": true,
  },
  "ok": true,
}
`;
